clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -325,4 +328,17 @@ void cleanup() {
             log.debug("Closed the consumer network thread");
         }
     }
+
+    /**
+     * If there is a metadata error, completed all uncompleted events with the metadata error.
+     */
+    private void maybeFailOnMetadataError(List<CompletableEvent<?>> events) {
+        if (networkClientDelegate.metadataError().isPresent()) {
+            Throwable metadataError = networkClientDelegate.metadataError().get();
+            if (!events.isEmpty()) {
+                events.forEach(event -> event.future().completeExceptionally(metadataError));
+                networkClientDelegate.clearMetadataError();
+            }
+        }
+    }
```

Comment:
Could we perform a pseudo `getAndClear()` with the metadata error?

```suggestion
    private void maybeFailOnMetadataError(List<CompletableEvent<?>> events) {
            if (events.isEmpty())
                return;

            Optional<Exception> error = networkClientDelegate.getAndClearMetadataError();
            error.ifPresent(e -> events.forEach(event -> event.future().completeExceptionally(e));
        }
    }
```